src/openfermion/utils/_sparse_tools.py

```diff
@@ -744,32 +744,34 @@ def expectation(operator, state):
     Raises:
         ValueError: Input state has invalid format.
     """
-
     if isinstance(state, scipy.sparse.spmatrix):
         # Handle density matrix.
         if isinstance(operator, scipy.sparse.linalg.LinearOperator):
             raise ValueError('Taking the expectation of a LinearOperator with '
                              'a density matrix is not supported.')
         product = state * operator
-        expectation = numpy.sum(product.diagonal())
+        return numpy.sum(product.diagonal())
 
-    elif isinstance(state, numpy.ndarray):
-        # Handle state vector.
+    if isinstance(state, numpy.ndarray):
         if len(state.shape) == 1:
             # Row vector
-            expectation = numpy.dot(numpy.conjugate(state), operator * state)
-        else:
-            # Column vector
-            expectation = numpy.dot(numpy.conjugate(state.T),
-                                    operator * state)[0, 0]
-
-    else:
-        # Handle exception.
-        raise ValueError(
-                'Input state must be a numpy array or a sparse matrix.')
-
-    # Return.
-    return expectation
+            return numpy.dot(numpy.conjugate(state), operator * state)
+        if len(state.shape) == 2:
+            if state.shape[1] == 1:
+                # Column vector
+                return numpy.dot(numpy.conjugate(state.T),
+                                 operator * state).item()
+            if state.shape[0] == state.shape[1]:
+                # Density matrix
+                product = state * operator
+                return numpy.sum(product.diagonal()).item()
```

Comment:
You can just use `numpy.trace`

Reply:
aha! you would think that. it does not work on sparse matrices

Reply:
Oh

Reply:
But multiplying a sparse matrix with a numpy array gives a numpy array. So here, `product` should be a numpy array, right?

Reply:
looks like this is correct.  state is a sparse matrix and operator is a sparse matrix.

Reply:
`state` is a Numpy array (see the if statement in line 755).